File: sdks/ruby/test/contract/compare.rb
# frozen_string_literal: true

module Contract
  module Compare
    module_function

    WEBHOOK_CODES = %w[
      invalid_signature
      timestamp_too_old
      malformed_signature
      missing_signature
      invalid_payload
    ].freeze

    def value_outcome(value)
      { ok: true, value: value }
    end

    def error_outcome(error)
      if error.is_a?(SolvaPay::PaywallError)
        return {
          ok: false,
          error_name: "PaywallError",
          error_message: error.message,
          error_kind: "Paywall",
        }
      end
      if error.is_a?(SolvaPay::SolvaPayError)
        kind = error.kind
        name = "SolvaPayError"
        kind = "Api" if kind.nil? && error.status
        kind = "Webhook" if WEBHOOK_CODES.include?(error.code)
        return {
          ok: false,
          error_name: name,
          error_message: error.message,
          error_status: error.status,
          error_kind: kind,
          error_code: error.code,
        }
      end
      {
        ok: false,
        error_name: error.class.name,
        error_message: error.message,
      }
    end

    def assert_expect(test, outcome, fixture)
      expect = fixture.fetch("expect")
      if expect.key?("result")
        test.assert outcome[:ok], "expected result, got #{outcome.inspect}"
        test.assert json_equal?(outcome[:value], expect["result"]),
                    "result mismatch:\n got: #{outcome[:value].inspect}\n expected: #{expect["result"].inspect}"
        return
      end

      expected = expect.fetch("error")
      test.refute outcome[:ok], "expected error, got #{outcome.inspect}"
      test.assert_equal expected.fetch("message"), outcome[:error_message], "error message"
      {
        "name" => :error_name,
        "status" => :error_status,
        "kind" => :error_kind,
        "code" => :error_code,
      }.each do |key, actual_key|
        test.assert_equal expected[key], outcome[actual_key], "error #{key}" if expected.key?(key)
      end
    end

    def json_equal?(left, right)
      if left.is_a?(Numeric) && right.is_a?(Numeric)
        return left.to_f == right.to_f
      end
      if left.is_a?(Array) && right.is_a?(Array)
        return left.length == right.length && left.zip(right).all? { |a, b| json_equal?(a, b) }
      end
      if left.is_a?(Hash) && right.is_a?(Hash)
        return left.keys.sort == right.keys.sort &&
               left.all? { |key, value| json_equal?(value, right[key]) }
      end
      left == right
    end
  end
end

File: sdks/python/tests/contract/compare.py
"""Outcome normalization and expectation comparison for the Python contract suite."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from solvapay.errors import PaywallError, SolvaPayError

from .fixture_loader import Fixture, FixtureErrorExpect


@dataclass(frozen=True)
class Outcome:
    ok: bool
    value: Any | None = None
    error_name: str | None = None
    error_message: str | None = None
    error_status: int | None = None
    error_kind: str | None = None
    error_code: str | None = None


def outcome_from_value(value: Any) -> Outcome:
    return Outcome(ok=True, value=value)


def outcome_from_exception(exc: BaseException) -> Outcome:
    if isinstance(exc, PaywallError):
        return Outcome(
            ok=False,
            error_name="PaywallError",
            error_message=str(exc),
            error_kind="Paywall",
        )
    if isinstance(exc, SolvaPayError):
        status = getattr(exc, "status", None)
        code = getattr(exc, "code", None)
        name = type(exc).__name__
        kind = "Api" if isinstance(status, int) else None
        if isinstance(code, str) and code in {
            "invalid_signature",
            "timestamp_too_old",
            "malformed_signature",
            "missing_signature",
            "invalid_payload",
        }:
            kind = "Webhook"
            name = "SolvaPayError"
        return Outcome(
            ok=False,
            error_name=name,
            error_message=str(exc),
            error_status=status if isinstance(status, int) else None,
            error_kind=kind,
            error_code=code if isinstance(code, str) else None,
        )
    return Outcome(
        ok=False,
        error_name=type(exc).__name__,
        error_message=str(exc),
    )


def json_eq(left: Any, right: Any) -> bool:
    """Deep equality with numeric coercion (`25` == `25.0`)."""
    if isinstance(left, (int, float)) and isinstance(right, (int, float)):
        return float(left) == float(right)
    if isinstance(left, list) and isinstance(right, list):
        return len(left) == len(right) and all(
            json_eq(a, b) for a, b in zip(left, right, strict=True)
        )
    if isinstance(left, dict) and isinstance(right, dict):
        if set(left) != set(right):
            return False
        return all(json_eq(left[key], right[key]) for key in left)
    return left == right


def assert_expect(outcome: Outcome, fixture: Fixture) -> None:
    if fixture.expect_kind == "result":
        assert outcome.ok, (
            f"expected result, got error "
            f"name={outcome.error_name!r} message={outcome.error_message!r}"
        )
        assert json_eq(outcome.value, fixture.expect_result), (
            f"result mismatch:\n  got:      {outcome.value!r}\n"
            f"  expected: {fixture.expect_result!r}"
        )
        return

    expected = fixture.expect_error
    assert expected is not None
    assert not outcome.ok, f"expected error {expected.message!r}, got success {outcome.value!r}"
    _assert_error(outcome, expected)


def _assert_error(actual: Outcome, expected: FixtureErrorExpect) -> None:
    assert actual.error_message == expected.message, (
        f"error message mismatch: got {actual.error_message!r}, expected {expected.message!r}"
    )
    if expected.name is not None:
        assert actual.error_name == expected.name, (
            f"error name mismatch: got {actual.error_name!r}, expected {expected.name!r}"
        )
    if expected.status is not None:
        assert actual.error_status == expected.status, (
            f"error status mismatch: got {actual.error_status!r}, expected {expected.status!r}"
        )
    if expected.kind is not None:
        assert actual.error_kind == expected.kind, (
            f"error kind mismatch: got {actual.error_kind!r}, expected {expected.kind!r}"
        )
    if expected.code is not None:
        assert actual.error_code == expected.code, (
            f"error code mismatch: got {actual.error_code!r}, expected {expected.code!r}"
        )
